blog/2023/06/PSA-dont-use-threadid.md

```diff
@@ -0,0 +1,176 @@
+
+# Fixing buggy code which uses this pattern
+
+## Quickfix: Replace `@threads` with `@threads :static`
+
+The simplest (though not recommended longterm) quickfix if you happen to use `@threads` is to replace usages of `@threads for ...` with `@threads :static for ...`. The reason for this is that the `:static` scheduler for `@threads` does not allow the asynchronous task migration and yielding that causes the bug in the first place.
+
+However, this is not a good long term solution because
+1) It's relying on guard rails to prevent otherwise incorrect code to be correct
+2) `@threads :static` is not cooperative or composable, that means that if code inside your `@threads :static` loop also does multithreading, your code could be reduced to worse than single-threaded speeds, or even deadlock.
+
+## Better fix: Work directly with tasks
+
+If you want a recipe that can replace the above buggy one with something that can be written using only the `Base.Threads` module, we recommend moving away from `@threads`, and instead working directly with `@spawn` to create and manage tasks. The reason is that `@threads` does not have any builtin mechanisms for managing and merging the results of work from different threads, whereas tasks can manage and return their own state in a safe way.
+
+Tasks creating and returning their own state is inherently safer than the spawner of parallel tasks setting up state for spawned tasks to read from and write to. 
+
+Code which replaces the incorrect code pattern shown above can look like this:
+```julia
+using Base.Threads: nthreads, @threads, @spawn
+using Base.Iterators: partition
+
+tasks_per_thread = 2 # customize this as needed. More tasks have more overhead, but better load balancing
+
+chunk_size = max(1, length(some_data) ÷ (tasks_per_thread * nthreads()))
+data_chunks = partition(some_data, chunk_size) # partition your data into chunks that individual tasks will deal with
+# See also ChunkSplitters.jl and SplittablesBase.jl for partitioning data
+
+tasks = map(data_chunks) do chunk
+    # Each chunk of your data gets its own spawned task that does its own local work and returns a result
+    @spawn begin
+        state = some_initial_value
+        for x in chunk
+            state = some_operator(state, f(x))
+        end
+        return state
+    end
+end
+states = fetch.(tasks) # get all the values returned by the individual tasks.
+# fetch is type unstable, so you may optionally want to assert a specific return value. 
+
+do_something(states)
+```
+
+This is a fully general replacement for the old, buggy pattern. However, for many applications this should be simplified down to a parallel version of `mapreduce`:
+```julia
+using Threads: nthreads, @spawn
+function tmapreduce(f, op, itr; init=some_initial_value, chunks_per_thread::Int = 2)
+    chunk_size = max(1, length(itr) ÷ chunks_per_thread)
```

Comment:
That again uses `nthreads`, which we explicitly want to get away from.

Reply:
> In my experience in most cases an estimation for a reasonable chunk size is easier at hand than the number
of chunks

Typically, I find the opposite to be true, but that's why I provided both options. In the version of the code that I wrote, if the user specifies `chunks_per_thread` then `chunk_size` will be calculated from that, but if they specify `chunk_size` then it'll use that `chunk_size` and the `chunks_per_thread` argument will be ignored.



> That again uses `nthreads`, which we explicitly want to get away from. It's not a reliable number to query, hence the inversion of responsibility, thinking about the amount of tasks for that work, instead of the amount of work per task.

No, this is a valid usage of `nthreads`. The correctness of the program does not depend on `nthreads` being reliable, it's just used here as a heuristic for how to split up work, which is fine.